compiler/infra/ForwardList.hpp

```diff
@@ -0,0 +1,684 @@
+#ifndef FORWARD_LIST_HPP
+#define FORWARD_LIST_HPP
+
+#pragma once
+
+#if 0
```

Comment:
Is this close to what you had in mind:
```
/**
 * @class TR::forward_list
 *
 * @brief A partial re-implementation of std::forward_list for C++03 supplemented by a linear-time size method.
 *
 * std::forward_list is a container introduced in the 2011 revision of the C++ standard for which
 * the objective is to provide the performance of a singly-linked list. Because the continer is singly-linked,
 * the interfaces common to other containers, where manipulations are performed on the half-closed interval
 * [begin, end), are problematic. To provide a work-around, std::forward_list provides similar operations that
 * operate on the open interval (before_begin, end), and provide [operation]_after variants that perform the
 * requested operation on the position immediately following the provided iterator.
 *
 * A further result of the strict requirement of being equally performant to a manually maintained singly-linked
 * list, is that std::forward_list does not provide a size method, as it cannot provide one that operates in O(1)
 * time. However, there are a number of places in our code base that depend on being able to request the size of
 * a container, where the container is one targetted for re-implementation as a forward_list. To reconcile these
 * requirements, we provide an extension to std::forward_list that provides a size() operation that runs in time
 * linear to the number of objects stored in the container.
 *
 */
```